pkg/importer/imageio-datasource.go

```diff
@@ -225,6 +236,258 @@ func (is *ImageioDataSource) IsDeltaCopy() bool {
+				return errors.Wrap(err, "failed to get range")
+			}
+			final := (index == (len(extentsReader.extents) - 1))
+			err = is.transferExtent(responseBody, outFile, extent, final)
+			if err != nil {
+				return errors.Wrap(err, "failed to transfer extent")
+			}
+		}
+	}
+
+	// A sync here seems to make it more likely that the next sync will work.
+	err = outFile.Sync()
+	if err != nil {
+		klog.Infof("Error from first attempt syncing %s: %v", fileName, err)
+	}
+
+	return nil
+}
+
+// transferExtent copies one extent from the source to the destination, updates the progress
+// counter, and closes the source. Each source reader is expected to contain one extent.
+func (is *ImageioDataSource) transferExtent(source io.ReadCloser, dest io.Writer, extent imageioExtent, final bool) error {
+	defer source.Close()
+	is.readers.progressReader.SetNextReader(source, final)
+
+	written, err := io.Copy(dest, is.readers.progressReader)
+	if err != nil {
+		return errors.Wrap(err, "failed to write to file")
+	}
+	if written != extent.Length {
+		return errors.New("failed to copy total extent length")
+	}
+
+	return nil
+}
+
+// monitorExtentsProgress sends a ticket renewal if there has been no download progress during the
+// polling time. This can happen if the destination storage does not have a fast way to punch holes.
+func (is *ImageioDataSource) monitorExtentsProgress(transferID string, extentsReader *extentReader, pollTime time.Duration, doneChannel chan struct{}) {
+	current := is.readers.progressReader.Current
+	for {
+		select {
+		case <-time.After(pollTime):
+			if is.readers.progressReader.Current <= current {
+				klog.Infof("No progress in the last %s, attempting ticket renewal to avoid timeout", pollTime)
+				if err := is.renewExtentsTicket(transferID, extentsReader); err != nil {
+					klog.Infof("Error renewing ticket: %v", err)
+				}
+			} else {
+				current = is.readers.progressReader.Current
+			}
+		case <-doneChannel:
+			klog.Info("Closing ticket expiration monitor")
+			return
+		}
+	}
+}
+
+// renewExtentsTicket ensures the ImageIO transfer ticket stays active by issuing a renewal and
+// by doing a small amount of I/O to prevent the oVirt engine from canceling the download.
+func (is *ImageioDataSource) renewExtentsTicket(transferID string, extentsReader *extentReader) error {
+	transfersService := is.connection.SystemService().ImageTransfersService()
+	transferService := transfersService.ImageTransferService(transferID)
+	_, err := transferService.Extend().Send()
+	if err != nil {
+		return errors.Wrap(err, "failed to renew transfer ticket")
+	}
+
+	readSize := 512
+	buf := make([]byte, readSize)
+	responseBody, err := extentsReader.GetRange(0, int64(readSize-1))
+	if err != nil {
+		return errors.Wrap(err, "failed sending small read to prevent download timeout")
+	}
+	defer responseBody.Close()
+
+	_, err = io.ReadFull(responseBody, buf)
+	return err
```

Comment:
No, it doesn't matter where this data goes as long as the server sees the I/O. Otherwise there's a good chance it will cancel the transfer ticket.

Reply:
So this is just dummy data, as long as we are showing activity. I just want to avoid having 512 byte gaps somewhere if we are supposed to write the data somewhere.

Reply:
This renewal always does its own independent range request on the first 512 bytes, so there shouldn't be any opportunity to cause a gap in the reading end or the writing end. For a short time I had this using extentReaders.Read, which did cause gaps because it was updating a file position offset that was shared with the main transfer. Calling GetRange directly makes sure that doesn't happen.

Reply:
Perfect, just want to make sure this was the case.